src/Storage/Parser.php

```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * This file is part of Lcobucci\JWT, a simple library to handle JWT and JWS
+ *
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ */
+
+declare(strict_types=1);
+
+namespace Lcobucci\JWT\Storage;
+
+use InvalidArgumentException;
+use Lcobucci\Jose\Parsing;
+use Lcobucci\JWT\Parser as ParserInterface;
+use Lcobucci\JWT\Token as TokenInterface;
+
+/**
+ * This class parses the JWT strings and convert them into tokens
+ *
+ * @author Luís Otávio Cobucci Oblonczyk <lcobucci@gmail.com>
+ * @since 0.1.0
+ */
+final class Parser implements ParserInterface
+{
+    /**
+     * The data decoder
+     *
+     * @var Parsing\Decoder
+     */
+    private $decoder;
+
+    /**
+     * Initializes the object
+     */
+    public function __construct(Parsing\Decoder $decoder)
+    {
+        $this->decoder = $decoder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function parse(string $jwt): TokenInterface
+    {
+        $data = $this->splitJwt($jwt);
+        $header = $this->parseHeader($data[0]);
+        $claims = $this->parseClaims($data[1]);
+        $signature = $this->parseSignature($header, $data[2]);
+
+        foreach ($claims as $name => $value) {
+            if (isset($header[$name])) {
+                $header[$name] = $value;
+            }
+        }
+
+        return new Token(
+            new DataSet($header, $data[0]),
+            new DataSet($claims, $data[1]),
+            $signature
+        );
+    }
+
+    /**
+     * Splits the JWT string into an array
+     *
+     * @param string $jwt
+     *
+     * @return array
+     *
+     * @throws InvalidArgumentException When JWT don't have all parts
+     */
+    private function splitJwt(string $jwt): array
+    {
+        $data = explode('.', $jwt);
+
+        if (count($data) != 3) {
+            throw new InvalidArgumentException('The JWT string must have two dots');
+        }
+
+        return $data;
+    }
+
+    /**
+     * Parses the header from a string
+     *
+     * @param string $data
+     *
+     * @return array
+     *
+     * @throws InvalidArgumentException When an invalid header is informed
+     */
+    private function parseHeader(string $data): array
+    {
+        $header = (array) $this->decoder->jsonDecode($this->decoder->base64UrlDecode($data));
+
+        if (isset($header['enc'])) {
+            throw new InvalidArgumentException('Encryption is not supported yet');
+        }
+
+        return $header;
+    }
+
+    /**
+     * Parses the claim set from a string
+     *
+     * @param string $data
+     *
+     * @return array
+     */
+    private function parseClaims(string $data): array
+    {
+        return (array) $this->decoder->jsonDecode($this->decoder->base64UrlDecode($data));
```

Comment:
It will return a recursive array however we would have a `TypeError` when an invalid claim set was given.